Source/DafnyRuntime/DafnyRuntime.py

```diff
@@ -107,22 +108,62 @@ def __lt__(self, other):
     def __le__(self, other):
         return len(self) <= len(other) and self == other[:len(self)]
 
-class Array(list):
-    @classmethod
-    def empty(cls, dims):
-        if dims <= 1:
-            return Array()
-        return Array([Array.empty(dims-1)])
+class Array:
+    class Box(list):
+        def __dafnystr__(self) -> str:
+            return '[' + ', '.join(map(string_of, self)) + ']'
+
+    def __init__(self, initValue, *dims):
+        fresh = not isinstance(initValue, Array.Box)
+        self.arr = initValue if fresh else initValue.content
+        self.dims = list(dims)
+        if fresh:
+            for i in reversed(self.dims):
+                self.arr = Array.Box([copy.deepcopy(self.arr) for _ in range(i)])
```

Comment:
I'm not sure the `deepcopy` is sound: won't it clone the elements of the array as well?

Reply:
I agree, we don't need this here, but I'm pretty sure it doesn't harm us either, as initialisations of reference types seem to be a two step process. First init to null and then override in a loop.